All Translog inner closes should be because tragedy exception has happened

andrershov opened the pull request:
We faced with the nasty race condition. See https://github.com/elastic/elasticsearch/issues/32526
`InternalEngine.failOnTragic` method has thrown `AssertionError`.
If you carefully look at if branches in this method, you will spot that its only possible, if either Lucene `IndexWriter `has closed from inside or `Translog`, has closed from inside, but tragedy exception is not set.
For now, let us concentrate on the `Translog` class.
We found out that there are two methods in Translog - namely `rollGeneration` and `trimOperations` that are closing Translog in case of Exception without tragedy exception being set.
This pull request fixes these to methods. To fix it, I've decided to pull `tragedyException` from `TranslogWriter` up-to `Translog` class.
Also to protect us in the future and make sure `close` method is never called from inside `Translog` special assertion examining stack trace is added. Since we're still targeting Java 8 for runtime - no StackWalker API is used in the implementation.

elasticmachine commented:
Pinging @elastic/es-distributed

bleskes commented:
@ywelsch please also discuss how far this can be backported (I only see labels for 7.0.0)

andrershov commented:
@ywelsch 

- I've wrapped AtomicReference with the tragic exception with TragedyExceptionHolder to make sure we do not lose, but instead suppress subsequent exceptions
- Made stack checking method static
- Added a test for stack checking method - the best thing we can do here is subclassing and check for subclasses in stack checking method
- Added non-null tragic exception assertions to tests that might fail during rollGeneration and trimOperations.

Could you please review my changes?

ywelsch commented on server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java:
```diff
@@ -2931,6 +2934,46 @@ public void testCloseSnapshotTwice() throws Exception {
     }
 
+    public void testTranslogCloseInvariant() throws IOException {
+        //close method should never be called directly from Translog (the only exception is closeOnTragicEvent)
```
also move this comment outside of the method

ywelsch replied:
finally, add an `assumeTrue("test only works with assertions enabled", Assertions.ENABLED);` to the beginning of the method

ywelsch commented on server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java:
```diff
@@ -129,28 +133,13 @@ public static TranslogWriter create(ShardId shardId, String translogUUID, long f
-            close();
-        } catch (final IOException | RuntimeException e) {
-            ex.addSuppressed(e);
+        if (tragedy.setTragicException(ex)) {
```
why change the semantics here to only call close when setting the tragedy the first time? Let's keep the existing semantics and make `setTragicException` return void

andrershov replied:
@ywelsch what is the reason to call close twice if every time we set tragic exception close is called?

ywelsch commented on server/src/main/java/org/elasticsearch/index/translog/TragicExceptionHolder.java:
```diff
@@ -0,0 +1,43 @@
+            tragedy.set(ex);
+            return true;
+        } else {
+            if (tragedy.get() != ex) { //to ensure there is no self-suppression
```
add space after `//`

ywelsch commented on server/src/main/java/org/elasticsearch/index/translog/TragicExceptionHolder.java:
```diff
@@ -0,0 +1,43 @@
+    public boolean setTragicException(Exception ex){
+        assert ex != null;
+        if (tragedy.compareAndSet(null, ex)) {
+            tragedy.set(ex);
```
why double-set here? Isn't the previous line already doing this?

andrershov commented:
@ywelsch thank you for the review!
I've fixed everything except close once. I've commented on it.

ywelsch commented on server/src/main/java/org/elasticsearch/index/translog/TragicExceptionHolder.java:
```diff
@@ -0,0 +1,43 @@
+     * Sets the tragic exception or if the tragic exception is already set adds passed exception as suppressed exception
+     * @param ex tragic exception to set
+     */
+    public void setTragicException(Exception ex){
```
space missing between `)` and `{`

ywelsch commented on server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java:
```diff
@@ -129,31 +133,16 @@ public static TranslogWriter create(ShardId shardId, String translogUUID, long f
         }
     }
 
+
```
remove extra newline

andrershov commented:
Merged to master, pending backport to 6.5.0 and 6.4.1
